### DISCUSSION TITLE
Proposal/Question: Location of Plugin Files in the Virtual FS

### DISCUSSION THREAD

**Discussion by JeffersGlass** [pyscript/pyscript#1229] [Proposals]
This proposal expands the section "Python Plugin Installation Location" from Discussion #1228 . See that discussion for context.

This is an unresolved question from the discussion; the location of the plugin files has implications for how users can use the plugin, and potentially implications for the behavior of imports within plugins themselves. **Feedback and further discussion is welcome and requested.**

# Where do Plugins Go in the Virtual Filesystem?

At the offsite, we didn't come to a conclusion as to where exactly the plugin entrypoint file (and any addtional files `fetch`'d by the plugin) end up in the Virtual Filesystem. For the sake of showing examples of the options we discussed, we'll invent a new `fado` plugin with a few files. We'll use the following partial metadata file:

```json
{
    "name": "fado",
    "language": "python",
    "entrypoint": "singer",
    "fetch": [
        {
            "files": ["singer.py"]
        },
        {
            "files": ["cod.py", "frenchfries.csv"],
            "to_folder": "dinner"
        },
    ]
}
```

## (1) Execution Directory
Files are fetched relative to the current working directory where Pyodide defaults to execution (/home/pyodide/), the same as they are for normal [[fetch]] configurations. The resulting file structure is:

```
/home/pyodide/
  ├─ pyscript.py
  ├─ singer.py
  ├─ dinner
  │  ├─ cod.py
  │  ├─ frenchfries.csv
```

This allows the user to write, in their \<py-script\> tags: 
  - `from singer import vasco`
  - `from dinner import cod`

## (2) Folder-Per-Plugin
For each plugin, a new folder is created in the home directory with the name of the plugin. The fetch configurations are executed relative to that folder. The resulting file structure is:

```
/home/pyodide/
  ├─ pyscript.py
  ├─ fado
  │  ├─ singer.py
  │  ├─ dinner
  │  │  ├─ cod.py
  │  │  ├─ frenchfries.csv
```

This allows the user to write, in their \<py-script\> tags: 
  - `from fado.singer import vasco`
  - `from fado.dinner import cod`

## (3) 'Plugins' Folder
A folder called `plugins` is created in the home directory. For each plugin a new folder is created within _that_ folder with the name of the plugin. The resulting file structure is:

```
/home/pyodide/
  ├─ pyscript.py
  ├─ plugins
  │  ├─ fado
  │  │  ├─ singer.py
  │  │  ├─ dinner
  │  │  │  ├─ cod.py
  │  │  │  ├─ frenchfries.csv
```

This allows the user to write, in their \<py-script\> tags: 
  - `from plugins.fado.singer import vasco`
  - `from plugins.fado.dinner import cod`

**Comment by antocuni** [comment]
#### (4) `plugins` folder, but we add it to `sys.path`

```
/home/pyodide/
  ├─ pyscript.py
  ├─ plugins
  │  ├─ fado
  │  │  ├─ singer.py
  │  │  ├─ dinner
  │  │  │  ├─ cod.py
  │  │  │  ├─ frenchfries.csv
```

Then at some point during the startup we do:
```python
sys.path.insert(0, '/home/pyodide/plugins')
```

And then we can do the following:
```python
from fado.singer import vasco
```
